### PR TITLE
React axe fixes

### PR DESCRIPTION
**Pull request by EdwinGuzman**
This partly resolves [SIMPLY-2385](https://jira.nypl.org/browse/SIMPLY-2385). Partly because I didn't want to make the PR too big and there are some issues that need more thought to fix.

Some that come to mind:
1) The SVG icons in the List page (the pyramids) all have the same ids. It may be better to just not have ids or pass in unique ids somehow.
2) Similar "duplicate id" issue for panels. If you go to the Troubleshooting page, for example, a lot of panels have ids of "undefined-panel". I thought it could be that we are simply not passing in an `id` prop. So, I tried passing in the name of the collection since that's what is being displayed, but then you have multiple instances of `axis-360-panel`. Even if you add an index, I kept seeing multiple `axis-360-0-panel`, `axis-360-1-panel`, etc. Any ideas would be helpful.
3) Axe keeps saying that the `ul.nav.nav-tabs` is not right since Bootstrap (it seems) adds `::before` and `::after` css pseudo elements. Not sure how to fix that unless if we step away from bootstrap but it will change a lot of the interface.

As for this PR, it depends on the [opds PR 242](https://github.com/NYPL-Simplified/opds-web-client/pull/242). So some issues are fixed there that pop up here. There's more information about which issues have been completed in the spreadsheet: https://docs.google.com/spreadsheets/d/1dWq1NLukCYJJx70id0V2wNst4Gjm_q6fRJZTRJUTVh0/edit#gid=1557754872
It's mostly adding aria-labels to forms, updating headings, and fixing some duplicate id issues.

### CODE REVIEW

File: src/components/BookEditForm.tsx [hunk 1]
```diff
@@ -44,13 +44,16 @@ export default class BookEditForm extends React.Component<BookEditFormProps, Boo
   }
 
   renderTextField(name: string, placeholder?: string, value?: string, hasLabel = true): JSX.Element {
+    const formattedName = formatString(name);
+    const ariaLabel = !hasLabel ? `Field for ${formattedName}` : null;
```
Comment (adriana-alter) [inline src/components/BookEditForm.tsx]: Why would there be a situation in which we would want the aria-label to be null?  Isn't `hasLabel` just about whether a label gets visually rendered?

File: src/components/ConfigPage.tsx [hunk 2]
```diff
@@ -33,14 +33,16 @@ export default class ConfigPage extends React.Component<ConfigPageProps, {}> {
     return (
       <div className="config">
         <Header />
```
Comment (adriana-alter) [inline src/components/ConfigPage.tsx]: I get why it's like this, but it seems bad to have `<main>` inside `<div>`.  Should the whole thing be the `<main>`, or would that be worse because the header and footer would be in there?

File: src/components/ClassificationsForm.tsx [hunk 3]
```diff
@@ -90,6 +90,7 @@ export default class ClassificationsForm extends React.Component<Classifications
                       type="text"
                       disabled={this.props.disabled}
                       name="target_age_min"
+                      aria-label="Target age minimum"
```
Comment (adriana-alter) [inline src/components/ClassificationsForm.tsx]: This and the other target age aria-label be in the imperative--i.e. "Enter a minimum target age"--for consistency with the aria-labels on input fields elsewhere.

File: src/components/ClassificationsTable.tsx [hunk 4]
```diff
@@ -22,8 +22,8 @@ export default class ClassificationsTable extends React.Component<Classification
             </tr>
           </thead>
           <tbody>
-            { this.props.classifications.map(classification =>
-              <tr key={classification.source + ":" + classification.name}>
+            { this.props.classifications.map((classification, idx) =>
+              <tr key={classification.source + "-" + idx}>
```
Comment (adriana-alter) [inline src/components/ClassificationsTable.tsx]: Why the change?  React specifically advises against using the idx as a key except as a last resort.  (I know the key here doesn't consist just of the idx...but still.)

File: src/components/CustomListEditor.tsx [hunk 5]
```diff
@@ -81,6 +81,8 @@ export default class CustomListEditor extends React.Component<CustomListEditorPr
               ) : null
             }
             <input
+              placeholder="Harry Potter"
```
Comment (adriana-alter) [inline src/components/CustomListEditor.tsx]: Did you mean to leave this in...?  (I mean, if we need a placeholder, this is a good one.  But it seems like a way to confuse admins.)

File: src/components/CustomListEntriesEditor.tsx [hunk 6]
```diff
@@ -190,17 +190,17 @@ export default class CustomListEntriesEditor extends React.Component<CustomListE
                 </div>
               }
             </div>
+            <p>Drag search results here to add them to the list.</p>
```
Comment (adriana-alter) [inline src/components/CustomListEntriesEditor.tsx]: Any particular reason for moving this?  (I don't have a problem with it; just curious.)

File: src/components/DiagnosticsServiceTabs.tsx [hunk 7]
```diff
@@ -34,14 +34,17 @@ export default class DiagnosticsServiceTabs extends TabContainer<DiagnosticsServ
     // If any of the timestamps has an exception, display a warning.  Otherwise, display the number of timestamps.
     let badge = hasException ?
         <span className="badge danger">!</span> :
-        <ToolTip trigger={<span className="badge">{timestampArray.length}</span>} text="Total number of timestamps for this service"/>;
+        <ToolTip
```
Comment (adriana-alter) [inline src/components/DiagnosticsServiceTabs.tsx]: Extra space :(

File: src/components/DiagnosticsServiceTabs.tsx [hunk 8]
```diff
@@ -34,14 +34,17 @@ export default class DiagnosticsServiceTabs extends TabContainer<DiagnosticsServ
...
+          text="Total number of timestamps for this service"
+        />;
 
     return <section><span className="service-name">{super.tabDisplayName(name)}</span>{badge}</section>;
   }
 
   renderCollections(collections: Array<DiagnosticsCollectionData>) {
     // If the collection has any timestamps with exceptions, it should start out expanded.
-    return Object.keys(collections).map((collectionName) =>
+    return Object.keys(collections).map((collectionName, idx) =>
```
Comment (adriana-alter) [inline src/components/DiagnosticsServiceTabs.tsx]: Previous question about using idx in keys holds for this too (and everywhere else it's happening!).

File: src/components/EditableInput.tsx [hunk 9]
```diff
@@ -38,75 +39,83 @@ export default class EditableInput extends React.Component<EditableInputProps, E
   }
 
   render() {
+    const {
+      type, elementType, optionalText, required, description, clientError, error,
+      label, extraContent,
+    } = this.props;
```
Comment (adriana-alter) [inline src/components/EditableInput.tsx]: Very nice!

File: src/components/GenreForm.tsx [hunk 10]
```diff
@@ -44,9 +44,11 @@ export default class GenreForm extends React.Component<GenreFormProps, GenreForm
         <div className="form-inline">
           <select
             name="genre"
+            aria-label="Select table for genres"
```
Comment (adriana-alter) [inline src/components/GenreForm.tsx]: This is kind of a confusing way to phrase that...

File: src/components/BookEditForm.tsx (hunk 1 above)
Comment (EdwinGuzman) [reply src/components/BookEditForm.tsx]: Yes, `hasLabel` is whether there is a visual label or not. If there is no visual label, the "fallback" would be to render an `aria-label` attribute. If it's null it just won't appear and that's okay because a visual label will be rendered.

File: src/components/SaveButton.tsx [hunk 11]
```diff
@@ -4,7 +4,7 @@ export interface SaveButtonProps {
   disabled: boolean;
   submit?: any;
   text?: string;
-  type?: string;
+  type?: "submit" | "reset" | "button";
 }
```
Comment (adriana-alter) [inline src/components/SaveButton.tsx]: Nice!

File: src/components/ClassificationsForm.tsx (hunk 3 above)
Comment (EdwinGuzman) [reply src/components/ClassificationsForm.tsx]: Good point.

File: src/components/Stats.tsx [hunk 12]
```diff
@@ -46,7 +46,7 @@ export class Stats extends React.Component<StatsProps, {}> {
     const totalStats = this.props.isLoaded && this.props.libraries && this.props.libraries.length > 1 && this.props.stats && this.props.stats["total"];
 
     return (
-      <div>
+      <>
```
Comment (adriana-alter) [inline src/components/Stats.tsx]: But...why?  Fragments seem less accessible...

File: src/components/TextWithEditMode.tsx [hunk 13]
```diff
@@ -7,6 +7,7 @@ export interface TextWithEditModeProps extends React.Props<TextWithEditMode> {
   text?: string;
   placeholder: string;
   onUpdate?: (text: string) => void;
+  "aria-label"?: string;
```
Comment (adriana-alter) [inline src/components/TextWithEditMode.tsx]: Why is this optional?

File: src/components/ClassificationsTable.tsx (hunk 4 above)
Comment (EdwinGuzman) [reply src/components/ClassificationsTable.tsx]: `classification.name` can be undefined. If it is, there can be multiple `${classification.source}:undefined` keys.

File: src/components/ConfigPage.tsx (hunk 2 above)
Comment (EdwinGuzman) [reply src/components/ConfigPage.tsx]: Right, the whole thing cannot be `<main>` because it shouldn't include the header or the footer. Technically it should be legal to do, but because there's only one header and because we don't intend to have sub-headers, then it will complain. I think a better change will be updating the whole app so that the header and footer always render and only the main content gets updated based on the route.

File: src/components/CustomListEditor.tsx (hunk 5 above)
Comment (EdwinGuzman) [reply src/components/CustomListEditor.tsx]: I may have thought it was a test. "Book title" is better. _deletrius_

File: src/components/CustomListEntriesEditor.tsx (hunk 6 above)
Comment (EdwinGuzman) [reply src/components/CustomListEntriesEditor.tsx]: `p` element shouldn't be in a `ul` element.

File: src/components/DiagnosticsServiceTabs.tsx (hunk 7 above)
Comment (EdwinGuzman) [reply src/components/DiagnosticsServiceTabs.tsx]: Hm?

File: src/components/DiagnosticsServiceTabs.tsx (hunk 8 above)
Comment (EdwinGuzman) [reply src/components/DiagnosticsServiceTabs.tsx]: Oh, I forgot to remove it. This is a case where the `Panel` element here needs to have an `id` prop. Currently, all the panels will have an id of `undefined-panel`. Using the index here doesn't help.

File: src/components/ClassificationsTable.tsx (hunk 4 above)
Comment (adriana-alter) [reply src/components/ClassificationsTable.tsx]: Ah.  That's bad.  How about telling it to use the name if there is one, and the idx if not?  Or is life too short for this?

File: src/components/EditableInput.tsx (hunk 9 above)
Comment (EdwinGuzman) [reply src/components/EditableInput.tsx]: This was actually accidental. Once I finished updating this, I realized I wanted to update a different function, but left it because it's slightly more readable.

File: src/components/GenreForm.tsx (hunk 10 above)
Comment (EdwinGuzman) [reply src/components/GenreForm.tsx]: True...

File: src/components/DiagnosticsServiceTabs.tsx (hunk 7 above)
Comment (adriana-alter) [reply src/components/DiagnosticsServiceTabs.tsx]: Blank line after line 37.  Or is Github just being weird?

File: src/components/SelfTestsCategory.tsx [hunk 14]
```diff
@@ -26,7 +26,18 @@ export default class SelfTestsCategory extends React.Component<SelfTestsCategory
     return (
       <div className="self-tests-category has-additional-content">
         <ul>
-          { this.props.items && this.props.items.map(i => <Panel style={getClassName(i)} key={i.name} openByDefault={onlyChild} headerText={i.name} content={[link(i), selfTests(i)]} />)}
+          { this.props.items && this.props.items.map((item, idx) =>
+              <li>
+                <Panel
+                  style={getClassName(item)}
+                  key={item.name}
+                  id={`${item.name.replace(/\s+/g, "")}-${idx}`}
```
Comment (EdwinGuzman) [inline src/components/SelfTestsCategory.tsx]: I should probably revert this as well and then fix all the Panel instances in one PR.

File: src/components/Timestamp.tsx [hunk 15]
```diff
@@ -28,6 +28,7 @@ export default class Timestamp extends React.Component<TimestampProps, {}> {
     // If the timestamp has an exception, it should start out expanded.
     return (
       <Panel
+        id="timestamp"
```
Comment (EdwinGuzman) [inline src/components/Timestamp.tsx]: But some `id` props in some instances are okay and won't be duplicated.

File: src/components/SelfTestsCategory.tsx (hunk 14 above)
Comment (adriana-alter) [reply src/components/SelfTestsCategory.tsx]: This sounds like a good idea.

File: src/components/ClassificationsTable.tsx (hunk 4 above)
Comment (EdwinGuzman) [reply src/components/ClassificationsTable.tsx]: Yeah, not bad to do.

File: src/components/DiagnosticsServiceTabs.tsx (hunk 7 above)
Comment (EdwinGuzman) [reply src/components/DiagnosticsServiceTabs.tsx]: Ah, that should be Github.

File: src/components/SaveButton.tsx (hunk 11 above)
Comment (EdwinGuzman) [reply src/components/SaveButton.tsx]: Yeah, I'm not sure why this issue didn't come up before 🤷‍♂  🤷‍♂

File: src/components/Stats.tsx (hunk 12 above)
Comment (EdwinGuzman) [reply src/components/Stats.tsx]: Fragments don't render anything. This is to reduce all the divs that get rendered. In this case, it doesn't add anything and seeing so many nested divs is confusing.

File: src/components/TextWithEditMode.tsx (hunk 13 above)
Comment (EdwinGuzman) [reply src/components/TextWithEditMode.tsx]: That can change. 👍

**Comment by EdwinGuzman** [comment]
I'm going to leave this open until other accessibility fixes are in to bundle them all together.
